rfd/0056-buildbox-image-handling.md

```diff
@@ -0,0 +1,171 @@
+
+#### 1. Immutable, independently-versioned buildboxes.
+
+That is:
+ * each buildbox image should be an immutable, versioned collection of tools required
+   to build Teleport. 
+ * The version of the overall  _collection_ should not be tied to the version of any
+   of the compenents inside it.
+ * Modifiying the contents of a buildbox implies creating a new version tag.
+
+#### 2. CI and releases are configured to use an _exact_ version of the buildbox
+
+CI & release proccesses refer to _exact_ versions of the buildbox they use, either 
+by tag (if we can guarantee tag immutability) or by hash (if we can't).
+
+ * This allows us to reproduce a build with the _exact same tooling_ that it was 
+   built with previously. We _cannot_ make this guarantee with floating tags
+ * We _certianly_ can't guarantee it with the build-at-time-of-use scheme currently
+   in use.
+ * Makes changes to the buildbox _opt-in_ for consuming tasks. For example: say 
+   somebody breaks the buildbox? No big deal - CI on your branch just keeps using the
+   old one until you're ready to upgrade.
+
+Ideally we should _also_ have some sort of developer-friendly floating tags so that
+people can experiment locally without having to sweat multiple release/update cycles.
+
+#### 3. Images being pushed to a repository with immutable tags
+
+This is in order to guarantee that the image is what we think it is. This is a developer
+affordance, as all this can be accomplished with hashes, but it's cumbersome and 
+unfriendly.
+
+#### 4. Images get built on a tag update and after code review & CI
+
+This imposes the same release mechanism on the tooling as on the main software. The idea 
+is to turn the process of spinning a new buildbox version into a rare(-r) event, rather 
+than happening on every merge.
+
+#### 5. Everything (including releases) uses the same set of images
+
+A release process should refernce an _exact_ buildbox image it will use and commit that
+to `git` prior to building the release. We should be able to see which buildbox was used
+for any given release just by examining the Teleport source.
+
+#### 6. Enforce strong validation of components on the buildbox as far as possible
+
+Currently we implicitly trust that _none_ of the following have been compromsed when
+constructing our buildbox:
+ - The Ubuntu debian package repositories
+ - The `go` package download package
+ - The `rustup` tool download package
+ - The `etcd` download package
+ - The Google `addlicense` repository
+ - The `libbpf` download site
+ - The Google Cloud Platform SDK download site
+ - The `golangci-lint` download site
+ - and a few others, but you get the idea...
+
+At minimum, we should create SHAs of known-good versions of the downloaded artefacts 
+and verify that they have not changed, to ensure that we are not surprosed by any 
+changes after the fact.
+
+#### 7. Buildbox source extracted from Teleport repo into it's own repo
+
+(This is controversial)
```

Comment:
Yeah, I'm not sure TBH we'd want to move it out but if the goal is to make it "less murky" can we move buildboxes out of build.assets into a separate "buildboxes" directory inside teleport repo with a proper readme/makefiles/etc. as you mention below?